src/VisualStudio/Core/Def/Implementation/PullMemberUp/MainDialog/PullMemberUpDialogViewModel.cs

```diff
@@ -0,0 +1,216 @@
+            _symbolToDependentsMap = dependentsMap;
+            Members = members;
+            _symbolToMemberViewMap = members.ToImmutableDictionary(memberViewModel => memberViewModel.Symbol);
+            _waitIndicator = waitIndicator;
+        }
+
+        public PullMembersUpOptions CreatePullMemberUpOptions()
+        {
+            var selectedOptionFromDialog = Members.
+                Where(memberSymbolView => memberSymbolView.IsChecked && memberSymbolView.IsCheckable).
+                SelectAsArray(memberViewModel =>
+                    (member: memberViewModel.Symbol,
+                    makeAbstract: memberViewModel.IsMakeAbstractCheckable && memberViewModel.MakeAbstract));
+
+            var options = PullMembersUpOptionsBuilder.BuildPullMembersUpOptions(
+                SelectedDestination.Symbol,
+                selectedOptionFromDialog);
+            return options;
+        }
+
+        public void SelectAllMembers()
+        {
+            SelectMembers(Members);
+        }
+
+        public void SelectPublicMembers()
+        {
+            SelectMembers(Members.WhereAsArray(memberViewModel => memberViewModel.Symbol.DeclaredAccessibility == Accessibility.Public));
+        }
+
+        public void EnableOrDisableOkButton()
+        {
+            var selectedMembers = Members
+                .WhereAsArray(memberSymbolView => memberSymbolView.IsChecked && memberSymbolView.IsCheckable);
+            OkButtonEnabled = SelectedDestination != null && selectedMembers.Count() != 0;
+        }
+
+        public void CheckAndSetStateOfSelectAllCheckBox()
+        {
+            var checkableMembers = Members.WhereAsArray(member => member.IsCheckable);
+            if (checkableMembers.All(member => member.IsChecked))
+            {
+                SelectAllCheckBoxState = true;
+                SelectAllCheckBoxThreeStateEnable = false;
+            }
+            else if (checkableMembers.Any(member => member.IsChecked))
+            {
+                SelectAllCheckBoxThreeStateEnable = true;
+                SelectAllCheckBoxState = null;
+            }
+            else
+            {
+                SelectAllCheckBoxState = false;
+                SelectAllCheckBoxThreeStateEnable = false;
+            }
+        }
+
+        public void SelectDependents()
+        {
+            var checkedMembers = Members
+                .WhereAsArray(member => member.IsChecked && member.IsCheckable);
+
+            var waitResult = _waitIndicator.Wait(
+                    title: ServicesVSResources.Pull_Members_Up,
+                    message: ServicesVSResources.Calculating_dependents,
+                    allowCancel: true,
+                    showProgress: true,
+                    context =>
+                    {
+                        foreach (var member in Members)
+                        {
+                            _symbolToDependentsMap[member.Symbol].Wait(context.CancellationToken);
```

Comment:
This is the place I feel a little bit complicated. So, since a member's dependents could be recursively and we just find its direct dependents in a new thread, we will need to wait for all work done before going to the Breadth-First search below otherwise it could be a race condition here (UI thread try to get an unfinished task).

I also have thought of trying to calculate the dependents recursively all in the background thread. So then it will become very interesting again since, during the calculation of one member we also need to calculate many other members' dependents, ideally, these results should all be cached to avoid calculate them twice, but since we are using Task.Run on each entry of the dictionary here so then maybe this dictionary should be locked .... 

So after some thoughts, I feel this is the easiest and safe way to do this temporary. I can continue to think how to do it and update this part later